### PR TITLE
add concurrency to NodeUpgradeJob

### PR DESCRIPTION
**Pull request by gy95**
**What type of PR is this?**

**What this PR does / why we need it**:
add concurrency to NodeUpgradeJob
support upgrading nodes one by one, but not upgrade them all once a time
**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: cloud/pkg/nodeupgradejobcontroller/controller/downstream.go [hunk 1]
```diff
@@ -162,65 +178,121 @@ func (dc *DownstreamController) nodeUpgradeJobAdded(upgrade *v1alpha1.NodeUpgrad
...
+		return
+	}
+}
+
+// selectConcurrentNodes select the nodes to do upgrade operation, and put it into channel nodesChan
+func (dc *DownstreamController) selectConcurrentNodes(nodesChan chan string, allNodes []string, upgrade *v1alpha1.NodeUpgradeJob) {
+	// the default concurrency is 1
+	// this means that we will upgrade nodes one by one
+	// only when the last one node upgrade finished, we'll continue to upgrade the next one node
+	concurrency := 1
```
Comment (wackxu) [inline cloud/pkg/nodeupgradejobcontroller/controller/downstream.go:247]: we can set the default value in the admission controller

File: cloud/pkg/nodeupgradejobcontroller/controller/downstream.go [hunk 2]
```diff
@@ -162,65 +178,121 @@ func (dc *DownstreamController) nodeUpgradeJobAdded(upgrade *v1alpha1.NodeUpgrad
...
+	if upgrade.Spec.Concurrency != 0 {
+		concurrency = int(upgrade.Spec.Concurrency)
+	}
+
+	timeout := int(*upgrade.Spec.TimeoutSeconds) * (len(allNodes) + 1) / concurrency
```
Comment (wackxu) [inline cloud/pkg/nodeupgradejobcontroller/controller/downstream.go:252]: Can you explain why we calculate timeout like this?

File: cloud/pkg/nodeupgradejobcontroller/controller/downstream.go [hunk 3]
```diff
@@ -162,65 +178,121 @@ func (dc *DownstreamController) nodeUpgradeJobAdded(upgrade *v1alpha1.NodeUpgrad
...
-		if err != nil {
-			klog.Warningf("marshal data failed: %v", err)
-			continue
+		// ensure the max number of upgrading nodes is Concurrency
+		if upgradingNum < concurrency {
+			for i := 0; i < concurrency-upgradingNum; i++ {
+				nodesChan <- allNodes[i]
+			}
+
+			allNodes = allNodes[:int(concurrency)-upgradingNum]
```
Comment (wackxu) [inline cloud/pkg/nodeupgradejobcontroller/controller/downstream.go:281]: allNodes = allNodes[int(concurrency)-upgradingNum:]？

Comment (gy95) [reply cloud/pkg/nodeupgradejobcontroller/controller/downstream.go:281]: no, line 278, we put allNodes[0], allNodes[1],...,allNodes[int(concurrency)-upgradingNum-1] node to nodesChan, so we should remove the heading `int(concurrency)-upgradingNum` nodes before continue the nexe loop

File: cloud/pkg/nodeupgradejobcontroller/controller/downstream.go (hunk 2 above)
Comment (gy95) [reply cloud/pkg/nodeupgradejobcontroller/controller/downstream.go:252]: use this just for ensure we can process all the timeout circumstance.

File: cloud/pkg/admissioncontroller/admission.go [hunk 4]
```diff
@@ -317,8 +321,44 @@ func (ac *AdmissionController) registerWebhooks(opt *options.AdmissionOptions, c
 		},
 	}
 
+	// mutatingWebhook contains all the kubeedge related Mutating webhook
+	mutatingWebhook := admissionregistrationv1.MutatingWebhookConfiguration{
```
Comment (fisherxu) [inline cloud/pkg/admissioncontroller/admission.go:325]: The name is better focus on the ability of the webhook, like `offlineMigrationWebhook` above

**Review by fisherxu — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**